File: src/tools/path-grants.ts
/**
 * Session-scoped out-of-workspace path grants.
 *
 * The two enforcement gates (file-tool `validatePathSafe` and the kernel bash
 * sandbox `defaultWritableRoots`) both default to "workspace only". This store
 * lets the agent — ONLY after an explicit user approval — widen that boundary
 * to a specific directory subtree, so authorized work outside the workspace
 * (writing a package to ~/Desktop, reading /tmp, touching the parent dir) is
 * possible without dropping the whole sandbox.
 *
 * Lifetime: grants live in-process (one session) by default. A grant may be
 * persisted per-workspace under ~/.rivet so a "remembered" path survives across
 * sessions of THAT workspace — never globally (a grant for project A must not
 * leak into project B).
 *
 * Security: a grant is a directory subtree. Containment checks canonicalize
 * symlinks on both sides so a granted path cannot be used to escape via a
 * symlinked child. `write` implies `read`.
 */
import { existsSync, mkdirSync, readFileSync, realpathSync } from 'node:fs'
import { homedir } from 'node:os'
import { join, resolve, sep } from 'node:path'
import { writeFileAtomicSync } from '../fs-atomic.js'
import { rivetHome } from '../config/paths.js'
import { expandHome } from '../platform.js'
import { debugLog } from '../utils/debug.js'
import { rawOutputDir } from './output-store.js'

export type GrantMode = 'read' | 'write'

export interface PathGrant {
  /** Canonicalized (realpath'd where possible) absolute directory root. */
  root: string
  mode: GrantMode
  grantedAt: number
  /** True when this grant was written through to the per-workspace store. */
  persisted?: boolean
  /**
   * Session scope: canonicalized cwd of the session that approved this grant.
   * Interactively approved grants (request_path_access) and per-workspace
   * persisted grants are ALWAYS scoped. Scope-less grants (config
   * permissions.additional*Dirs, dependency-cache read grants, rivet runtime
   * dirs) are user- or tool-level by design and visible to every session in
   * the process.
   *
   * The sidecar (`rivet serve`) hosts many sessions — possibly across
   * different workspaces — in ONE process, so an unscoped interactive grant
   * approved in workspace A would silently authorize writes from workspace B.
   * The header contract below ("a grant for project A must not leak into
   * project B") is enforced through this field.
   */
  scope?: string
}

const RIVET_DIR = rivetHome()

/** In-memory grants for the current process. Interactive/persisted grants are
 *  session-scoped via PathGrant.scope; scope-less entries are user-level. */
let _grants: PathGrant[] = []

/** True when grant `g` is visible to the session running in `cwd`. An omitted
 *  cwd (legacy callers) sees everything. */
function scopeVisible(g: PathGrant, cwd?: string): boolean {
  if (!cwd) return true
  if (g.scope === undefined) return true
  return g.scope === canonicalize(cwd)
}

/**
 * Canonicalize a path: resolve symlinks where the path (or its nearest existing
 * ancestor) exists, so containment checks compare real paths. Falls back to a
 * plain resolve for not-yet-existing targets.
 */
function canonicalize(p: string): string {
  const abs = resolve(p)
  try {
    return realpathSync(abs)
  } catch {
    // Walk up to the nearest existing ancestor, canonicalize it, re-append tail.
    let current = abs
    const tail: string[] = []
    while (!existsSync(current)) {
      const parent = resolve(current, '..')
      if (parent === current) return abs // reached fs root
      tail.unshift(current.slice(parent.length + 1))
      current = parent
    }
    try {
      return join(realpathSync(current), ...tail)
    } catch {
      return abs
    }
  }
}

/**
 * Windows filesystems are case-insensitive: drive letters and path segments
 * arrive in mixed case (F:\ vs f:\, cmd vs Explorer casing), and realpath only
 * normalizes casing for path components that exist. Containment/dedup checks
 * must therefore fold case on win32 or grants silently fail to match.
 */
const CASE_INSENSITIVE_FS = process.platform === 'win32'

function foldCase(p: string): string {
  return CASE_INSENSITIVE_FS ? p.toLowerCase() : p
}

/**
 * True when `child` is the same as `root` or nested under it, using a
 * separator boundary so `/a/b` does NOT match `/a/bc`. Exposed with an
 * explicit case-sensitivity flag for unit testing win32 semantics on any host.
 *
 * win32 accepts `/` and `\` interchangeably, so both fold to the host
 * separator before the boundary check — a mixed-separator path can neither
 * slip past nor falsely fail containment (posix-style test data must work on
 * a win32 host too). On posix `\` is a legal filename character and stays
 * literal.
 */
export function isPathUnder(root: string, child: string, caseInsensitive: boolean = CASE_INSENSITIVE_FS): boolean {
  const fold = (p: string): string => {
    const norm = sep === '\\' ? p.replace(/\//g, '\\') : p
    return caseInsensitive ? norm.toLowerCase() : norm
  }
  const r = fold(root)
  const c = fold(child)
  if (c === r) return true
  const prefix = r.endsWith(sep) ? r : r + sep
  return c.startsWith(prefix)
}

function isUnder(root: string, child: string): boolean {
  return isPathUnder(root, child)
}

/** Per-workspace persisted-grants file, keyed by a cwd slug (mirrors checkpoint.ts). */
function grantsFile(cwd: string): string {
  const slug = canonicalize(cwd).replace(/[^a-zA-Z0-9]/g, '_').slice(-64)
  return join(RIVET_DIR, `path-grants-${slug}.json`)
}

/**
 * Grant access to a directory subtree. `root` is canonicalized. A write grant
 * supersedes a prior read grant on the same root. When `opts.persist` is set,
 * the grant is also written to the per-workspace store (requires opts.cwd).
 *
 * When `opts.cwd` is provided the grant is SCOPED to that workspace/session:
 * it authorizes only sessions running in the same cwd. Interactively approved
 * grants must always pass cwd — an unscoped grant would leak across every
 * session the sidecar hosts. Callers that intentionally grant process-wide
 * (config permissions, dependency caches, rivet runtime dirs) omit cwd.
 */
export function grantPath(root: string, mode: GrantMode, opts?: { persist?: boolean; cwd?: string }): PathGrant {
  const canonical = canonicalize(root)
  const scope = opts?.cwd ? canonicalize(opts.cwd) : undefined
  const persist = opts?.persist === true
  const existing = _grants.find(g =>
    foldCase(g.root) === foldCase(canonical)
    && g.scope === scope,
  )
  let grant: PathGrant
  if (existing) {
    // Upgrade read → write; never downgrade.
    if (mode === 'write') existing.mode = 'write'
    if (persist) existing.persisted = true
    grant = existing
  } else {
    grant = { root: canonical, mode, grantedAt: Date.now(), ...(persist ? { persisted: true } : {}), ...(scope ? { scope } : {}) }
    _grants.push(grant)
  }
  if (persist && opts?.cwd) persistGrants(opts.cwd)
  return grant
}

/** True if `absPath` is under any granted root visible to `cwd` (read or write satisfies read). */
export function isReadGranted(absPath: string, cwd?: string): boolean {
  const target = canonicalize(absPath)
  return _grants.some(g => scopeVisible(g, cwd) && isUnder(g.root, target))
}

/** True if `absPath` is under any WRITE-granted root visible to `cwd`. */
export function isWriteGranted(absPath: string, cwd?: string): boolean {
  const target = canonicalize(absPath)
  return _grants.some(g => g.mode === 'write' && scopeVisible(g, cwd) && isUnder(g.root, target))
}

/** Write-granted roots visible to `cwd` (consumed by the sandbox's writable-roots builder). */
export function writeGrantedRoots(cwd?: string): string[] {
  return _grants.filter(g => g.mode === 'write' && scopeVisible(g, cwd)).map(g => g.root)
}

/** Snapshot of current grants. */
export function listGrants(): PathGrant[] {
  return _grants.map(g => ({ ...g }))
}

/** Read the per-workspace store, tolerating a missing/corrupt file. */
function readPersistedFile(cwd: string): PathGrant[] {
  const file = grantsFile(cwd)
  if (!existsSync(file)) return []
  try {
    const parsed = JSON.parse(readFileSync(file, 'utf-8')) as PathGrant[]
    if (!Array.isArray(parsed)) return []
    return parsed.filter((g): g is PathGrant => !!g && typeof g.root === 'string')
  } catch {
    return []
  }
}

/**
 * The grants persisted for this workspace, as stored on disk. Display/revocation
 * surfaces read this rather than `listGrants()`: the in-memory store also holds
 * session-only grants and the dependency/runtime read grants, which the user
 * never authorized explicitly and must not be offered as revocable entries.
 */
export function listPersistedGrants(cwd: string): PathGrant[] {
  return readPersistedFile(cwd).map(g => ({
    root: g.root,
    mode: g.mode === 'write' ? 'write' : 'read',
    grantedAt: typeof g.grantedAt === 'number' ? g.grantedAt : 0,
    persisted: true,
  }))
}

/**
 * Revoke a grant by exact root. Both halves are required: dropping only the file
 * would leave the subtree writable for the rest of the session (a "revoked"
 * grant that still writes), and dropping only memory would resurrect it at the
 * next startup via `loadPersistedGrants`.
 *
 * Matching is exact, not containment — revoking `/a` must not silently remove a
 * separately-granted `/a/b` the user never asked about.
 *
 * The file is rewritten from its own contents rather than from memory: a peer
 * session may have persisted a grant after this process hydrated, and rewriting
 * from our (staler) memory would silently drop it.
 */
export function revokeGrant(root: string, opts: { cwd: string }): boolean {
  const canonical = canonicalize(root)
  const matches = (candidate: string): boolean => foldCase(canonicalize(candidate)) === foldCase(canonical)
  // Only this workspace's scoped grants (plus scope-less user-level grants) are
  // revocable here — another workspace's in-memory grant must not be touched.
  const revocable = (g: PathGrant): boolean => matches(g.root) && (g.scope === undefined || g.scope === canonicalize(opts.cwd))

  const hadInMemory = _grants.some(revocable)
  if (hadInMemory) _grants = _grants.filter(g => !revocable(g))

  const onDisk = readPersistedFile(opts.cwd)
  const kept = onDisk.filter(g => !matches(g.root))
  const removedFromDisk = kept.length < onDisk.length
  if (removedFromDisk) {
    try {
      mkdirSync(RIVET_DIR, { recursive: true })
      writeFileAtomicSync(grantsFile(opts.cwd), JSON.stringify(kept, null, 2))
    } catch {
      /* best-effort: the in-memory revocation already took effect this session */
    }
  }
  return hadInMemory || removedFromDisk
}

/** Write the currently-persisted grants to the per-workspace store. */
function persistGrants(cwd: string): void {
  try {
    mkdirSync(RIVET_DIR, { recursive: true })
    // 只写本工作区作用域的持久授权——sidecar 单进程多会话共享 _grants，
    // 不过滤会把 B 区 hydrated 的授权写进 A 的文件，A 下次启动即以 scope=A
    // 注水（收编 PR #84 时补齐的泄漏面，评审遗漏接线点③）。
    const scope = canonicalize(cwd)
    const toSave = _grants.filter(g => g.persisted && g.scope === scope)
    writeFileAtomicSync(grantsFile(cwd), JSON.stringify(toSave, null, 2))
  } catch {
    /* best-effort: a persistence failure must not break the grant itself */
  }
}

/**
 * Hydrate persisted grants for this workspace into the in-memory store at
 * startup. Re-canonicalizes each root (paths may have moved/symlinks changed)
 * and drops any whose root no longer exists.
 */
export function loadPersistedGrants(cwd: string): void {
  const file = grantsFile(cwd)
  if (!existsSync(file)) return
  let saved: PathGrant[]
  try {
    saved = JSON.parse(readFileSync(file, 'utf-8')) as PathGrant[]
  } catch {
    return
  }
  if (!Array.isArray(saved)) return
  for (const g of saved) {
    if (!g || typeof g.root !== 'string') continue
    if (!existsSync(g.root)) continue
    const mode: GrantMode = g.mode === 'write' ? 'write' : 'read'
    // Scoped to this workspace: another workspace's persisted grants must not
    // become visible to sessions running here (sidecar hosts many cwds).
    grantPath(g.root, mode, { persist: false, cwd })
    const stored = _grants.find(x =>
      foldCase(x.root) === foldCase(canonicalize(g.root))
      && x.scope === canonicalize(cwd),
    )
    if (stored) stored.persisted = true
  }
}

/**
 * 盘根/目录存在性探测的 TTL 记忆（2026-09-05 跨盘审批链顺手项）：
 * applyConfiguredPathGrants 对每条配置目录同步 existsSync——Windows 上死映射
 * 盘符/断连网络盘的单次探测可达秒级，单线程 sidecar 上 26 个盘根逐个探测会
 * 冻结事件循环（审批事件都发不出去）。启动与 PUT /config/permission-dirs 都
 * 会全量重放，同一批路径短时间内反复探测纯属浪费。
 *
 * 两条路径两档 TTL（2026-09-22 复查修正——此前共用 30s 让 GET 稳态形同未修）：
 *  - **授权应用**（applyConfiguredPathGrants）30s：授权要尽快跟随磁盘真值；
 *    用户刚保存的路径另有 forceRoots 当场实测，不吃记忆。
 *  - **显示探针**（probeConfiguredDirExists，GET/PUT 响应的 `exists` 字段）5min：
 *    桌面端 usePermissionDirs 的 staleTime 是 60s，TTL 若短于它，稳态下每次 GET
 *    到达时记忆必然已过期 → 照样逐盘裸 existsSync（26 盘根 × 秒级 = 冻事件循环）。
 */
const ROOT_EXISTS_TTL_MS = 30_000
const PROBE_EXISTS_TTL_MS = 5 * 60_000
const rootExistsMemo = new Map<string, { exists: boolean; at: number }>()

export function resetRootExistsMemoForTest(): void {
  rootExistsMemo.clear()
}

function rootExistsCached(root: string, force = false, ttlMs: number = ROOT_EXISTS_TTL_MS): boolean {
  const now = Date.now()
  if (!force) {
    const hit = rootExistsMemo.get(root)
    if (hit && now - hit.at < ttlMs) return hit.exists
  }
  const exists = existsSync(root)
  rootExistsMemo.set(root, { exists, at: now })
  return exists
}

/**
 * 配置目录的"存在么"探测（GET/PUT /config/permission-dirs 的 `exists` 字段）。
 *
 * 必须走 TTL 记忆，不能裸 existsSync：桌面端「全盘只读」在 Windows 上会写入
 * 26 个盘根，而该 GET 路由每次 AutonomyMenu 挂载（60s stale 之后）都会打一次，
 * 逐个同步探测 = 单线程 sidecar 冻结事件循环（审批事件都发不出去，UI 整体
 * "卡住"）。
 *
 * TTL=5min 而非 30s：前端 staleTime 60s，稳态下 30s 记忆必过期，GET 仍会全量
 * 重探——修了等于没修。代价是 exists 字段最多滞后 5 分钟（设置页 typo/缺失盘
 * 提示足够；新保存的路径经 PUT 的 forceRoots 当场实测，不吃记忆）。
 */
export function probeConfiguredDirExists(raw: string): boolean {
  const trimmed = raw.trim()
  if (!trimmed) return false
  return rootExistsCached(resolve(expandHome(trimmed)), false, PROBE_EXISTS_TTL_MS)
}

/**
 * Apply the user's standing directory grants from config
 * (`permissions.additionalReadDirs` / `additionalWriteDirs`) at session start —
 * the Codex-style "give this folder to the agent" model. Session-scoped
 * in-memory grants (config is the durable source; nothing is written to the
 * per-workspace grant store). Non-existent entries are skipped fail-closed:
 * a typo'd config line must not open a subtree that later comes into being.
 *
 * 探测成本（2026-09-22 修复）：`force`（"用户刚保存的路径当场实测"）此前是整批
 * 语义——桌面上「全盘只读」一次写入 26 个盘根，于是 save 一次就强制同步探测
 * 26 次，Windows 死映射盘/断连网络盘单次可达秒级 → 事件循环冻结。现在只有
 * **本次新增**的路径强制实测（`forceRoots`），未变路径走 30s TTL 记忆：新挂载
 * 的盘照样当场可见，重复保存不再堵住事件循环。
 */
export function applyConfiguredPathGrants(
  permissions: { additionalReadDirs?: string[]; additionalWriteDirs?: string[] } | undefined,
  opts?: { force?: boolean; forceRoots?: readonly string[] },
): void {
  if (!permissions) return
  const forced = new Set((opts?.forceRoots ?? [])
    .map(r => r.trim())
    .filter(Boolean)
    .map(r => resolve(expandHome(r))))
  const apply = (dirs: string[] | undefined, mode: GrantMode): void => {
    for (const raw of dirs ?? []) {
      const trimmed = raw.trim()
      if (!trimmed) continue
      const root = resolve(expandHome(trimmed))
      if (!rootExistsCached(root, opts?.force === true || forced.has(root))) continue
      grantPath(root, mode, { persist: false })
    }
  }
  apply(permissions.additionalReadDirs, 'read')
  apply(permissions.additionalWriteDirs, 'write')
}

/**
 * Common third-party dependency / toolchain read-only cache directories
 * (relative to $HOME). Read-side counterpart of the writable roots enumerated
 * in `sandbox-profile.ts::defaultWritableRoots` and `sandbox-toolchain.ts` —
 * the same directories the bash sandbox already lets commands write to.
 *
 * Reading a project's dependency source (a HarmonyOS fork in `.pub-cache`, a
 * git-sourced package, a version-mismatched transitive dep) is a routine
 * diagnostic step. Without these read grants, `read_file` / `grep` on any path
 * under $HOME trips `validatePathSafe` → an approval `await` that has no
 * timeout, and the batch-time watchdog is disarmed (turn-orchestrator.ts:892),
 * so the session hangs until the user hits Ctrl+C — exactly the failure seen
 * in the kaiyang session reading `~/.pub-cache/git/.../*.dart`.
 *
 * Grants are READ-ONLY, session-scoped, never persisted: write still goes
 * through the normal approval/sandbox path, and nothing leaks across projects
 * or sessions. Non-existent entries are skipped fail-closed.
 */
const DEFAULT_DEPENDENCY_READ_DIRS = [
  '.npm', '.npm-cache', '.cache',
  '.cargo', '.rustup',
  '.gradle', '.m2',
  '.pub-cache',
  '.pnpm-store', '.yarn',
  '.bun', '.deno',
  '.cocoapods',
  '.gem', '.bundle',
  '.composer', '.config/composer',
  '.android',
  '.nuget',       // .NET NuGet packages
  '.nvm',         // Node Version Manager installations
  '.pyenv',       // Python version manager installs
  'go',
  // In-project node_modules / vendor / Pods already live under cwd and need no
  // grant; this list only covers $HOME-level global caches.
]

/**
 * Environment variable overrides for dependency cache directories.
 *
 * Each key matches a DEFAULT_DEPENDENCY_READ_DIRS entry. When the named
 * environment variable is set (non-empty), its value is used as the absolute
 * directory root instead of the default `$HOME/{key}` path. Absent or empty
 * env vars fall through to the default.
 *
 * Only well-known, user-facing env vars with a stable contract are listed
 * (e.g. CARGO_HOME, GRADLE_USER_HOME). Npm/yarn/pnpm use layered config
 * (npmrc, yarnrc) whose env-var surface is unreliable; they stay on defaults.
 *
 * See .rivet/knowledge/debug-t7-collapse-cache-cliff.md §4 — env var is the
 * write side (who sets), not the read side (who reads).
 */
const ENV_OVERRIDES: Record<string, string> = {
  '.cargo': 'CARGO_HOME',
  '.rustup': 'RUSTUP_HOME',
  '.gradle': 'GRADLE_USER_HOME',
  '.pub-cache': 'PUB_CACHE',
  '.gem': 'GEM_HOME',
}

export function applyDefaultDependencyReadGrants(): void {
  const home = homedir()
  const granted: string[] = []
  const skipped: string[] = []
  for (const rel of DEFAULT_DEPENDENCY_READ_DIRS) {
    const envVar = ENV_OVERRIDES[rel]
    const root = envVar && process.env[envVar]
      ? resolve(process.env[envVar])
      : resolve(join(home, rel))
    if (!existsSync(root)) {
      skipped.push(rel)
      continue
    }
    grantPath(root, 'read', { persist: false })
    granted.push(rel)
  }
  debugLog(`dep read grants: ${granted.length} granted` +
    (granted.length ? ` (${granted.join(', ')})` : '') +
    (skipped.length ? `; ${skipped.length} skipped (${skipped.join(', ')})` : ''))
}

/**
 * Read grants for directories Rivet itself writes outside the workspace and
 * then tells the model to read back.
 *
 * `$TMPDIR/rivet-raw` holds the full output of tools whose result was truncated
 * for the model, and the truncation footer says verbatim:
 * `full output: read_file <rawPath> — 不要重跑命令`. Without a grant that
 * `read_file` fails `validatePathSafe` ("Path outside project directory"), so
 * the model can neither recover the output nor re-run the command — a closed
 * dead end (9 occurrences across 4 sessions on 2026-07-27/28).
 *
 * Unlike the dependency caches above these roots are NOT existence-gated: the
 * raw dir is created lazily on the first truncated output, so at session start
 * it usually does not exist yet and an `existsSync` skip would make the grant
 * never take effect. Skipping fail-closed is right for user-configured paths
 * (a typo must not open a subtree) but wrong here — the path is ours, derived
 * from code rather than input, and it only ever contains output this agent
 * itself produced.
 *
 * Read-only, session-scoped, never persisted.
 */
export function applyRivetRuntimeReadGrants(): void {
  const roots = [rawOutputDir()]
  for (const root of roots) {
    grantPath(root, 'read', { persist: false })
  }
  debugLog(`rivet runtime read grants: ${roots.join(', ')}`)
}

/** Test-only: clear the in-memory grant store. */
export function _resetGrantsForTest(): void {
  _grants = []
}

File: go/internal/agent/pathgrants.go
// pathgrants.go —— 运行时路径授权存储。
//
// 对账 TS `src/tools/path-grants.ts`（506 行）的**核心授予/查询语义**。
//
// ## 为什么需要它
//
// 审批决策树的 `pathGrantNeed` 分支有两种处理，都需要运行时授权：
//
//   - **skip 档**（`dangerously-skip-permissions`）→ 出界路径**首触即授**
//     （对账 TS `tool-pipeline.ts:1240-1249`）
//   - 其他档 → 弹审批；批准后 `grantPath` 记录授权，后续同路径不再问
//
// Go 侧此前只有 `StaticGrantChecker`（静态、构造期固定）——无法在运行时
// 授予，故 skip 档的"零打断"语义无法实现。
//
// ## scope 收窄（明示）
//
// TS 的完整能力含**持久化**（`persistGrants` / `loadPersistedGrants`，
// 写 `<rivetHome>/path-grants-<slug>.json`）。本文件**只实现进程内存储**——
// 持久化留待需要时补（当前 skip 档的会话级授予足够；TS 注释也说
// 「会话级，不出会话」）。
//
// 未移植：`listPersistedGrants` / `loadPersistedGrants` / `persistGrants` /
// `revokeGrant` / `applyConfiguredPathGrants` / `applyDefaultDependencyReadGrants` /
// `applyRivetRuntimeReadGrants` / `probeConfiguredDirExists`。这些服务于
// config 装配与依赖缓存读取面，各自有独立消费者。
//
// ## 安全语义（逐条对账 TS，不可简化）
//
//  1. **canonicalize**：解析符号链接（最近存在祖先），防符号链接绕过
//  2. **case folding**：Windows 文件系统大小写不敏感，`F:\` 与 `f:\` 必须等价
//  3. **段边界**：`/a/b-evil` 不得匹配 `/a/b`（防 fail-open）
//  4. **scope 隔离**：授权绑定 session cwd——A 工作区的授权不得泄漏给 B
//  5. **写覆盖读**：同 root 的 write 授予升级 read 授予，**永不降级**
package agent

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"

	"github.com/kalandramo/tianshu/go/internal/pathsafe"
)

// GrantMode 是授权模式（对账 TS `GrantMode`）。
type GrantMode string

const (
	GrantRead  GrantMode = "read"
	GrantWrite GrantMode = "write"
)

// PathGrant 是一条路径授权（对账 TS `PathGrant`）。
//
// **未含 TS 的 `persisted` / `grantedAt` 字段**——持久化未移植，
// `grantedAt` 无消费者。需要时补。
type PathGrant struct {
	// Root 是规范化（realpath）后的绝对目录根。
	Root string
	// Mode 是授权模式。
	Mode GrantMode
	// Scope 是会话作用域（规范化后的 session cwd）。
	//
	// **为什么必须有**：sidecar 在一个进程里承载多个会话（可能跨不同
	// 工作区），无作用域的交互式授权会让工作区 A 的批准静默授权工作区 B
	// 的写操作。空串 = 进程级（配置/依赖缓存类，由用户或工具级授予）。
	Scope string
}

// grantModeFromPathsafe 把 pathsafe 的访问模式转为授权模式。
//
// **为什么需要转换**：两个包的 Mode 是独立类型（`pathsafe.Mode` 用 iota 枚举，
// `GrantMode` 用字符串常量——后者对账 TS 的 `'read' | 'write'` 字面量类型）。
// 合并成一个会破坏各自的对账锚（TS 侧本就是两套）。
func grantModeFromPathsafe(m pathsafe.Mode) GrantMode {
	if m == pathsafe.ModeWrite {
		return GrantWrite
	}
	return GrantRead
}

// caseInsensitiveFS 报告当前平台文件系统是否大小写不敏感。
//
// 对账 TS `CASE_INSENSITIVE_FS`（Windows / macOS 默认不敏感）。
func caseInsensitiveFS() bool {
	return runtime.GOOS == "windows" || runtime.GOOS == "darwin"
}

// foldCase 按平台语义折叠大小写（对账 TS `foldCase`）。
func foldCase(p string) string {
	if caseInsensitiveFS() {
		return strings.ToLower(p)
	}
	return p
}

// isPathUnder 报告 child 是否位于 root 之下（**按路径段边界**）。
//
// 对账 TS `isPathUnder`。两处语义不可简化：
//
//   - **段边界**：`/a/b-evil` 不得匹配 `/a/b`——`strings.HasPrefix` 会
//     fail-open（项目已记录该坑，`StaticGrantChecker` 的 `hasPathPrefix`
//     同源）。
//   - **分隔符归一**：Windows 上 `/` 与 `\` 等价。
func isPathUnder(root, child string) bool {
	norm := func(p string) string {
		if runtime.GOOS == "windows" {
			p = strings.ReplaceAll(p, "/", `\`)
		}
		return foldCase(p)
	}
	r := norm(root)
	c := norm(child)
	if c == r {
		return true
	}
	sep := string(filepath.Separator)
	prefix := r
	if !strings.HasSuffix(prefix, sep) {
		prefix += sep
	}
	return strings.HasPrefix(c, prefix)
}

// canonicalizePath 规范化路径：解析符号链接（或最近存在祖先），使包含性
// 检查比较的是**真实路径**。
//
// 对账 TS `canonicalize`。**安全关键**：不解析符号链接的话，攻击者可经
// 符号链接把"工作区内"指向工作区外。
//
// 目标不存在时向上走到最近的存在祖先，规范化它再接回尾部——这样
// 「将要创建的文件」也能得到正确判定。
func canonicalizePath(p string) string {
	abs, err := filepath.Abs(p)
	if err != nil {
		abs = filepath.Clean(p)
	}
	if real, err := filepath.EvalSymlinks(abs); err == nil {
		return real
	}
	// 向上找最近存在祖先
	current := abs
	var tail []string
	for {
		if _, err := os.Lstat(current); err == nil {
			break
		}
		parent := filepath.Dir(current)
		if parent == current {
			return abs // 到根仍未找到
		}
		tail = append([]string{filepath.Base(current)}, tail...)
		current = parent
	}
	if real, err := filepath.EvalSymlinks(current); err == nil {
		parts := append([]string{real}, tail...)
		return filepath.Join(parts...)
	}
	return abs
}

// pathGrantStore 是进程内授权存储。
//
// 对账 TS 的包级 `_grants` 数组 + 各查询函数。用 mutex 保护——sidecar
// 多会话并发访问。
type pathGrantStore struct {
	mu     sync.RWMutex
	grants []PathGrant
}

// newPathGrantStore 构造空的授权存储。
func newPathGrantStore() *pathGrantStore {
	return &pathGrantStore{}
}

// scopeVisible 报告授权 g 对 cwd 会话是否可见（对账 TS `scopeVisible`）。
//
// 空 cwd 视为「看得到全部」（legacy 调用方）；无 scope 的授权是进程级的。
func (s *pathGrantStore) scopeVisible(g PathGrant, cwd string) bool {
	if cwd == "" {
		return true
	}
	if g.Scope == "" {
		return true
	}
	return g.Scope == canonicalizePath(cwd)
}

// GrantPath 授予目录子树的访问权（对账 TS `grantPath`）。
//
// `root` 被规范化。**写授权覆盖先前的读授权**（永不降级）。`cwd` 非空时
// 授权**绑定该工作区**——交互式批准必须传 cwd，否则会跨会话泄漏。
func (s *pathGrantStore) GrantPath(root string, mode GrantMode, cwd string) PathGrant {
	canonical := canonicalizePath(root)
	scope := ""
	if cwd != "" {
		scope = canonicalizePath(cwd)
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	for i := range s.grants {
		g := &s.grants[i]
		if foldCase(g.Root) == foldCase(canonical) && g.Scope == scope {
			// 升级 read → write，永不降级。
			if mode == GrantWrite {
				g.Mode = GrantWrite
			}
			return *g
		}
	}
	grant := PathGrant{Root: canonical, Mode: mode, Scope: scope}
	s.grants = append(s.grants, grant)
	return grant
}

// IsReadGranted 报告路径是否在任一可见授权根之下（对账 TS `isReadGranted`）。
//
// 读授权或写授权都满足读（写蕴含读）。
func (s *pathGrantStore) IsReadGranted(absPath, cwd string) bool {
	target := canonicalizePath(absPath)
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, g := range s.grants {
		if s.scopeVisible(g, cwd) && isPathUnder(g.Root, target) {
			return true
		}
	}
	return false
}

// IsWriteGranted 报告路径是否在任一可见的**写**授权根之下（对账 TS `isWriteGranted`）。
func (s *pathGrantStore) IsWriteGranted(absPath, cwd string) bool {
	target := canonicalizePath(absPath)
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, g := range s.grants {
		if g.Mode == GrantWrite && s.scopeVisible(g, cwd) && isPathUnder(g.Root, target) {
			return true
		}
	}
	return false
}

// ListGrants 返回当前授权的快照（对账 TS `listGrants`）。
func (s *pathGrantStore) ListGrants() []PathGrant {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]PathGrant, len(s.grants))
	copy(out, s.grants)
	return out
}

// ResetGrantsForTest 清空授权（测试用，对账 TS `_resetGrantsForTest`）。
func (s *pathGrantStore) ResetGrantsForTest() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.grants = nil
}
